[2단계 - 페이먼츠] 클레어(안은소) 미션 제출합니다.

eunsoA opened the pull request:
## 🎯 페이먼츠

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계
- 재사용 가능한 Input Component를 개발한다.
- Storybook을 사용하여 컴포넌트의 다양한 상태를 시각적으로 테스트한다.
- 카드 정보를 효과적으로 렌더링 하기 위한 상태 관리를 경험한다.

### 2단계
- 다양한 Form 구성 요소들간의 상태를 효율적으로 관리한다.
- hooks API를 이용하여 상태 관리 로직을 구현하한다.
- custom hooks를 생성하여 Form 관리 로직을 컴포넌트에서 분리하고 재사용한다.
- Controlled & Uncontrolled Components에 입각하여 Form을 핸들링한다.

---

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항(코드 컨벤션, 에러 핸들링 등)을 준수했나요?
- [x] 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 기본 배포 링크 기입: https://eunsoa.github.io/react-payments/

<br/>

### 리뷰 요청 & 논의하고 싶은 내용

안녕하세요 ☀️해먼드 ~! 7기 클레어입니다!

우선 결혼 다시 한번 축하드려요 ~! 🥳🎊
신혼여행으로 바쁘실텐데 지난 step1 에서 꼼꼼히 봐주신 덕분에 많이 배우고 생각해볼 수 있었어요. 감사해요 ! 

<br/>

#### 🔽 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점
1. config 사용
  - 알고 봤더니 config를 사용한 사람이 이번 미션에서 저와 다른 한분 총 2명 뿐이더라구요. 사실 과정 속에서 불편함이 많았지만, 이번 미션은 이걸 끝까지 가져가면서 불편함을 직접 느껴보고자, step2에서도 config 구조를 유지한 채로 구현했어요. (해먼드 말대로 `과도한 공통화`를 통해 적정한 기준선을 찾기 위함이었어요 ㅎㅎ)
  - config 구조를 갖고 있음으로 오는 장단점이 있는 것 같아요. 같은 type에 대해서는 섹션을 추가하기 쉬웠지만, 다른 타입의 섹션의 경우, 유지보수하기 상당히 골치가 아파지는.. 문제가 있더라구요. 게다가 다른 타입으로 인해서 `CardInfoSection`에서는 depth가 분리되는데,, 상당히 보기 좋지 않아서 왜 다들 안쓰는지..! 이해했습니다! ㅎㅎ
  - 만약 다른 사람과 함께 해야했다면 config를 사용하지 않았을 것 같아요. 그러나 step2는 혼자 진행하는 만큼 config를 그대로 사용하면서 불편함을 느껴보고자했어요. (그리고 이제 충분히 느낀 것 같아서 리팩터링하면서 수정하고싶어요.) 
  - 해먼드의 생각은 어떤가요?

2. FSD 아키텍처
  - 이번 기회에 처음 써보게 되어서 step2에서도 이 아키텍처에 맞춰 폴더 구조를 분리하고, 관심사를 분리하고자 했어요.
  - FSD 아키텍처의 핵심 철학은 '기능 중심 구조'로, 코드를 비즈니스 기능 단위로 분리하여 폴더 단위로 협업하기 좋은 구조라고 이해했어요. 쉽게 말해서 협업시에 각자 하나의 폴더를 맡고, 각 폴더가 책임이 명확하게 분리되어 있어서 협업하기 좋은 구조인거죠. 
  - 이런 관점에서 봤을때 현재 제 설계 구조는 이 목적에 부합하지 않는 것 같아요.  features 하위에 크게 `cardRegistration(카드 등록 페이지)`, `cardComplete(카드 등록 완료 페이지)`, `preview`, `cardInfo` 이렇게 분리되고 있어요. 페이지로 분리해보면 `preview`, `cardInfo` 는 모두 `cardRegistration`에 포함되는 것 같아요. 그럼 `preview`, `cardInfo` 는 `cardRegistration`로 들어가는게 맞을까요? ( 개인적으로 아직은 명확한 책임 분리에 따른 관심사 분리..!에 대해 명확한 주관을 갖기 어려운 것 같아요. )
    ```
    📦src
     ┣ 📜main.tsx
     ┣ 📜index.css
     ┣ 📜vite-env.d.ts
     ┣ 📂app            👈 app, router와 관련된 로직
     ┃ ┣ 📜routes.ts
     ┃ ┗ 📜App.tsx
     ┣ 📂shared      
     ┃ ┣ 📂ui           👈 공통적으로 사용되는 로직
     ┃ ┣ 📂constants
     ┃ ┗ 📂lib
     ┣ 📂features       👈 ui 로직
     ┃ ┣ 📂cardRegistration
     ┃ ┣ 📂cardComplete
     ┃ ┣ 📂preview
     ┃ ┗ 📂cardInfo
     ┣ 📂entities       👈 비즈니스 로직
     ┃ ┗ 📂cardInfo
     ┗ 📂stories
    ```


#### 🔽 2) 이번 리뷰를 통해 논의하고 싶은 부분

1. 상태 관리
- 지금 `useCardInfo`라는 하나의 훅으로 모든 상태가 관리되고 있어요. 저는 카드 정보라는 하나의 개념(훅)으로 묶여서 관리되는 게 이상적이라고 생각해서 초기에 이런 구조로 만들었어요.
- 아무래도 이런 구조이다 보니 `features/cardInfo/hooks/useCardInfo`가 굉장히 무거워요. 이걸 어떤 식으로 분리해야할지 막막한데, 각 카드 정보에 대해서 상태를 분리해야할지 고민이 돼요.
- 이에 대한 해먼드의 생각이 궁금해요!


2. 가독성이 좋은 코드인가?
- 협업의 관점에서 보았을때, 읽기 쉬운 코드가 가장 좋은 코드라고 생각해요. 그리고 그러한 코드를 짜고 싶어요. 
- 그런면에서 과도한 공통화는 추상화로 인해 읽기 어려워지고 제 목적과 멀어진다는 생각이 들었어요. config 관련한 질문과도 이어지는 것 같은데요, 이에 대한 해먼드의 생각이 궁금해요. 
- 좀 더 가독성 좋은 코드로 리팩터링하려면 어떤 방향성을 잡으면 좋을지 해먼드의 생각이 궁금해요.


#### 🔽 데모 영상

https://github.com/user-attachments/assets/c14467cf-f6e5-469c-acc3-d2c5030d9c35

<br/>
최대한 깔끔하게 코드를 쓰고 싶었는데, 다 완성하고 보니 참 부끄럽고 읽기 어려운 코드가 되어 버린 것 같아서 아쉽네요 ㅠㅠ 🥺 많은 채찍질과 피드백 환영이예요! 그럼, 이번 step2도 잘 부탁드립니다! 🙇‍♂️
좋은 하루 보내세요 :) 🍀
<br/><br/>

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도만이 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. Form 상태 관리 & Custom Hook 분리

- 반복되는 로직을 custom hook으로 분리했는가?
- hook 내부와 UI 컴포넌트의 **역할이 명확하게 분리**되어 있는가?
- 상태 흐름이 직관적이며, Form 전체를 일관되게 관리할 수 있는 구조인가?

### 2. 입력 UI 흐름과 UX

- 카드 번호 입력 시 필드 간 **자동 포커싱 이동**이 자연스럽게 동작하는가?
- 숫자만 입력 가능한 필드에서 제한, 에러 메시지, 유효성 피드백 등 **사용자 경험이 충분히 고려**되었는가?

### 3. 컴포넌트 구조 및 재사용성

- 컴포넌트가 명확한 **역할과 책임**을 가지며 과도하게 분리되거나 중첩되어 있지 않은가?
- Input, Button 등 재사용 가능한 컴포넌트가 잘 정의되어 있는가?

### 4. 상태 기반 유효성 검사 및 확인 버튼 활성화

- 모든 필드가 유효할 때만 확인 버튼이 **정확히 활성화/비활성화**되는가?
- 유효성 검사의 기준이 명확하고, **상태 변경에 따른 UI 반응**이 잘 연결되어 있는가?

eunsoA committed:
refactor: cardType 상수에서 FIRST_DIGIT 값을 문자열에서 숫자로 수정

eunsoA committed:
refactor: 범위 체크를 위한 isInRange 유틸 함수를 shared/lib 하위로 분리

eunsoA committed:
refactor: cardInfoValidator을 명령형에서 선언형으로 수정

eunsoA committed:
refactor: cardInfoSectionConfig에 유효성 검사 및 에러키 추가

eunsoA committed:
refactor: 에러 메시지와 상태값을 cardErrorConstants.ts로 상수화

eunsoA committed:
feat: CardInfoType 상수 추가 및 cardInfoSectionConfig에서 사용

eunsoA committed:
style: cardInfoSection의 에러 메세지 높이 고정 속성 추가

eunsoA committed:
refactor: useCardInfo 훅에서 카드 정보 유효성 검사 로직을 분리

eunsoA committed:
fix: key의 prefix 삭제

eunsoA committed:
docs: step2 기능 요구 사항 추가

eunsoA committed:
refactor: CARD_INFO_VALID_RULE 추가하여 cardInfoSectionConfig의 maxLength 설정

eunsoA committed:
feat: 카드 정보 입력 섹션에 대한 동적 UI 기능 추가

eunsoA committed:
style: Preview UI의 border-radius 및 box-shadow 속성 수정

eunsoA committed:
feat: CVC 유효성 검사에 3자리 체크 추가 및 카드 번호, 만료일 유효성 검사 로직 개선

eunsoA committed:
chore: react-router-dom 및 관련 의존성 추가

eunsoA committed:
refactor: CARD_IMAGE_PATH 위치 이동 및 import 경로 수정

eunsoA committed:
feat: 체크 아이콘 SVG 추가 및 ICON_PATH 상수 정의

eunsoA committed:
refactor: ERROR_MESSAGES를 Object.freeze로 변경하여 불변성 보장

eunsoA committed:
feat: CardRegistration, CardComplete 페이지 추가 및 라우팅 설정

eunsoA committed:
feat: CardRegistration 폼 제출 처리 변경

eunsoA committed:
feat: CardComplete 페이지에서 카드 번호 동적 표시 기능 추가

eunsoA committed:
feat: 카드 정보 입력 섹션에 비밀번호 필드 추가 및 유효성 검사 로직 구현

eunsoA committed:
feat: icon_arrow_down SVG 추가 및 ICON_PATH에 경로 정의

eunsoA committed:
style: select 요소 스타일링 초기화

eunsoA committed:
refactor: cardType 관련 로직 Utils로 분리

eunsoA committed:
feat: CARD_COMPANIES 상수 및 색상 정의 추가

eunsoA committed:
feat: 카드 정보 입력 섹션에 카드사(cardCompany) 선택 필드 추가 및 관련 로직 구현

eunsoA committed:
style(cardInfoSection): px에서 rem으로 변경

eunsoA committed:
feat: 카드사 관련 유틸리티 함수 추가

eunsoA committed:
feat: 카드사에 따른 배경색 동적 적용 및 CSS 수정

eunsoA committed:
feat: CardCplete 페이지에서 카드사 정보 동적 표시 기능 추가

eunsoA committed:
feat: 카드 유효성 검사 로직에 현재 월 및 연도 이후 검사 추가

eunsoA committed:
fix: 카드 유효성 검사 로직에서 연도 및 월 비교 조건 수정

eunsoA committed:
style: Preview와 CardRegistration 섹션의 submit 버튼 및 스크롤 CSS 수정

eunsoA committed:
fix: 카드 유효성 검사 로직에서 월과 연도 비교 조건 수정

eunsoA committed:
fix: 개발 환경에 따라 routes 동적 설정

eunsoA committed:
style: cardInfoContainer CSS에서 margin 및 padding 값 수정

eunsoA committed:
style: select 요소의 아이콘 이미지 경로 수정

JunilHwang commented on src/entities/cardInfo/model/cardInfoValidator.ts:
```diff
@@ -1,70 +1,140 @@
+import {
+  CARD_INFO_VALID_RULE,
+  CardInfoType,
+} from '../../../entities/cardInfo/constants/cardInfoTypeConstants';
```
import 경로가 불필요하게 긴 것 같네요..!?

`../constants/cardInfoTypeConstants` 이렇게 표현할 수 있지 않을까요?

eunsoA replied:
어랏 이 부분 수정했습니다! 감사합니다! (commit : 17d5b1279377bc0db1b34ad3ab41abeb5a17d839)

JunilHwang commented on src/entities/cardInfo/utils/cardCompanyUtils.ts:
```diff
@@ -0,0 +1,23 @@
+  CARD_COMPANIES_DEFAULT_COLOR,
+} from '../constants/cardCompanyConstants';
+
+export const getCardCompanyKey = (companyName: string): keyof typeof CARD_COMPANIES | undefined => {
```
`keyof typeof CARD_COMPANIES` 를 미리 정의해주면 코드가 조금 더 깔끔해질 것 같아요..!

```tsx
type CardCompanyKey =  keyof typeof CARD_COMPANIES;
```

요로코롬..

eunsoA replied:
오 좋습니다! 이 내용도 반영하여서 수정했어요. (commit : 273c3559cad65fa73275b83416b9577982ce3797)
더 깔끔하네용 ㅎㅎ

JunilHwang commented on src/shared/ui/cardInfoSelect.css:
```diff
@@ -0,0 +1,22 @@
+  border-radius: 2px;
+  border:
+    1.015px solid,
+    #acacac;
```
CSS 문법 오류가 있네요..!
`1.015px soild #acacac;` 처럼 표현해야 한답니다!

eunsoA replied:
앗 이부분도 수정했습니다! 감사해요 :) (commit : bfa9fed2498e28f92ed94e89fb102a6a8dc9eb29)

JunilHwang commented on src/features/cardComplete/ui/CardComplete.tsx:
```diff
@@ -0,0 +1,34 @@
+import { useNavigate, useLocation } from 'react-router-dom';
+import { ROUTES } from '../../../app/routes';
```
features가 app을 의존하고 있네요..!?

eunsoA replied:
> 일단 feature에서 app을 의존하는 부분이 있어요. 의존성이 꼬여버린거죠 ㅎㅎ 이런 부분을 우선적으로 개선해주시면 좋겠어요. 
> 그리고 FSD를 따라하는 방법도 좋고, 개인적으로 FSD에서 어떤 방식으로 관심사를 관리하고자 하는지를 분석해보면 좋다고 생각해요. 가령, feature, entities, shared 같은 계층을 꼭 만들어야 하는건 아니라고 생각해요. 
> 비슷한 유형의 의존성을 묶고, 의존성의 방향이 단방향으로 흐르도록 만들어주기만 해도 충분하지 않나 싶습니다.

말씀해주신 것처럼 의존성의 방향이 단방향으로 흐르지 않던 문제들이 있더군요!

  - FSD 아키텍처의 기본 레이어 
      - app: 앱 초기화, 라우팅, 글로벌 설정
      - processes (선택적): 여러 feature를 조합한 상위 프로세스
      - pages (선택적): 라우팅 단위 화면
      - features: 특정 기능(유저 액션 중심)의 UI/로직
      - entities: 비즈니스 도메인 모델과 관련된 로직/데이터
      - shared: 재사용 가능한 UI, 유틸, 상수 등
      👉 의존성 방향: shared → entities → features → processes → pages → app

위의 기본적인 FSD 아키텍처 구조에서 저는 아래의 순서로 의존성 방향을 잡고, Feature/ 에서 app/ 을 의존하는 문제도 해결했습니다!
>  `Shared` -> `entities` -> `features` -> `pages` -> `app` 

(Commit : [27433e2](https://github.com/woowacourse/react-payments/pull/475/commits/27433e29461f72f45c11102d0dbe68469c1f9c50), [53b156f](https://github.com/woowacourse/react-payments/pull/475/commits/53b156f8e3d498fa34967bcd40c8773563b24990))

JunilHwang commented on src/features/cardInfo/hooks/useCardInfo.tsx:
```diff
@@ -1,64 +1,107 @@
+  SECTION_ORDER,
+} from '../../../entities/cardInfo/constants/cardInfoTypeConstants';
 
 export default function useCardInfo() {
```
일단 useCardInfo 내부 코드가 200줄이 넘어가고 있어요! 그렇다면 더 많은 hook으로 관심사를 분리할 수 있지 않을까요!?

eunsoA replied:
네 동의해요!! 아래와 같은 구조로 분리하였습니다!

- `useCardNumber`, `useCardCompany`, `useCardExpDate`, `useCardCVC`, `useCardPassword`: 각 카드 정보 타입별 처리 로직
- `useSectionManager`: 섹션 상태 및 네비게이션 관리
- `useErrorHandler`: 에러 상태 관리
- `useCardInfo` : 모든 개별 훅을 조합하여 통합된 카드 정보 관리

(commit : 8c79f0be05bb608fb2d951e134beed52dccd6962)

JunilHwang commented on src/features/cardInfo/hooks/useCardInfo.tsx:
```diff
@@ -1,64 +1,107 @@
+    [CardInfoType.EXPDATE]: { month: '', year: '' },
+    [CardInfoType.CVC]: '',
+    [CardInfoType.PASSWORD]: '',
   });
```
여기서 쓰이는 기본 값의 경우 리액트 상태에 대한 의존성이 없기 때문에 훅 밖으로 분리할 수 있지 않을까 싶네요..!

eunsoA replied:
이전에 useCardInfo에서 모든 상태가 관리될때는 훅 밖으로 분리할 수 있었을텐데, 현재는 아래와 같은 구조로 의존성 때문에 훅 밖으로 분리하기 어려울 것 같아요!

```ts
const { cardNumber, changeCardNumber } = useCardNumber( ... );

  const { cardCompany, changeCardCompany } = useCardCompany( ... );

  const { cardExpirationDate, changeCardExpDate } = useCardExpDate( ... );

  const { cardCVC, changeCardCVC } = useCardCVC( ... );

  const { cardPassword, changeCardPassword } = useCardPassword( ... );

  const cardInfo: CardInfo = {
    [CardInfoType.NUMBER]: cardNumber,
    [CardInfoType.COMPANY]: cardCompany,
    [CardInfoType.EXPDATE]: cardExpirationDate,
    [CardInfoType.CVC]: cardCVC,
    [CardInfoType.PASSWORD]: cardPassword,
  };
```

( 관련 변경사안 commit : 8c79f0be05bb608fb2d951e134beed52dccd6962)

JunilHwang commented on src/features/cardInfo/hooks/useCardInfo.tsx:
```diff
@@ -1,64 +1,107 @@
   });
 
-  const handleCardInfoChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+  const handleCardNumberChange = (index: number, value: string) => {
```
cardNumber의 값을 변경하는 "기능"을  수행하고 있기 때문에, 이벤트 핸들러라기보단 하나의 함수라고 볼 수 있지 않을까요?

그래서 이름을 handle~ 로 쓰는게 아니라

`changeCardNumber` 처럼 표현하는게 더 자연스럽다고 생각해요!

eunsoA replied:
오 다시 보니 그렇네요..! 반영하여서 수정했어요! :) 감사합니다!
(commit : 5c2c1ba52ce9fba52e5273e163403019b3386b59)

JunilHwang commented on src/features/cardInfo/hooks/useCardInfo.tsx:
```diff
@@ -1,64 +1,107 @@
+    });
+  };
+
+  const handleCardCompanyChange = (value: string) => {
```
앞선 내용과 동일합니다!

eunsoA replied:
동일하게 수정했습니다! (commit : 5c2c1ba52ce9fba52e5273e163403019b3386b59)

JunilHwang commented on src/features/cardInfo/hooks/useCardInfo.tsx:
```diff
@@ -1,64 +1,107 @@
+  const handleCardNumberChange = (index: number, value: string) => {
+    setCardInfo((prev) => {
+      const updatedNumbers = prev.cardNumber.map((num, i) => (i === index ? value : num));
+      validateCardNumber(updatedNumbers);
```
set 함수 내부에서 다시 함수를 하나 호출하는데 이게 또 set을 하고 있네요..!
저는 validate 라는 이름 때문에 검증 후에 오류를 발생시키는 생각했는데 예상과는 다르군요 ㅎㅎ

eunsoA replied:
으앗 다시 보니 그렇네요..!
useCardInfo 내부가 비대해지다보니 각 함수별로 역할과 책임이 명확히 분리되지 않았던 것 같아요.
아래와 같이 각 함수의 역할과 책임을 분리했습니다!

  1. 상태 업데이트 함수 (update-)
    - 데이터만 업데이트하는 순수 함수로, 각 카드 정보 타입별로 분리
  2. 데이터 검증 함수 (validate-)
    - 입력 데이터 검증 후 boolean 값(isValid) 반환
    - 에러 상태 업데이트 담당
  3. 완료 여부 체크 함수 (is-Complete)
    - 각 섹션의 입력 완료 여부만 체크하는 순수 함수
  4. 섹션 관리 함수 (setSectionState, completeSection)
    - 섹션 상태 업데이트(setSectionState) 및 다음 섹션으로 이동 처리(completeSection)
  5. 핸들러 함수 (change-)
    - 상태 업데이트, 검증, 완료 여부 확인을 순차적으로 처리
  6. 이벤트 라우팅 함수 (handleCardInfo)
    - 각 카드 정보 타입별 입력을 올바른 핸들러로 라우팅
  
(commit : b1021e500f94c9c474a3817b72fd7181d8b9d94d )

JunilHwang commented on src/features/cardInfo/hooks/useCardInfo.tsx:
```diff
@@ -1,64 +1,107 @@
+    });
+  };
+
+  const handleCardInfoChange = (e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>) => {
```
이게 진짜 handler고 나머지는 함수의 조합인 것 같아요.
그마저도 reducer 같은걸 만들어서 관리할 수 있지 않나 싶어요!

eunsoA replied:
오 reducer를 사용하는 건 생각 안해봤었는데요, 아래와 같이 변경하는 걸 말씀하는게 맞으신가요?! 생각보다 더 복잡해보여서요 ... 🥲

```ts
// 변경 전
const handleCardInfo = (e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>) => {
    const { name, value } = e.target;

    if (name.startsWith(CardInfoType.NUMBER)) {
      const index = Number(name[name.length - 1]);
      changeCardNumber(index, value);
      return;
    }

    if (name === CardInfoType.COMPANY) {
      changeCardCompany(value);
      return;
    }

    if (name.startsWith(CardInfoType.EXPDATE)) {
      const key = name.split('-')[1] as 'month' | 'year';
      changeCardExpDate(key, value);
      return;
    }

    if (name.startsWith(CardInfoType.CVC)) {
      changeCardCVC(value);
      return;
    }

    if (name.startsWith(CardInfoType.PASSWORD)) {
      changeCardPassword(value);
      return;
    }
  };
```
```ts
// reducer 사용
const cardInfoReducer = (action: CardAction) => {
    switch (action.type) {
      case CARD_ACTIONS.CHANGE_CARD_NUMBER:
        const { index, value } = action.payload;
        changeCardNumber(index, value);
        break;
      case CARD_ACTIONS.CHANGE_CARD_COMPANY:
        changeCardCompany(action.payload);
        break;
      case CARD_ACTIONS.CHANGE_CARD_EXP_DATE:
        const { key, value: expDateValue } = action.payload;
        changeCardExpDate(key, expDateValue);
        break;
      case CARD_ACTIONS.CHANGE_CARD_CVC:
        changeCardCVC(action.payload);
        break;
      case CARD_ACTIONS.CHANGE_CARD_PASSWORD:
        changeCardPassword(action.payload);
        break;
    }
  };

  const actionMap = {
    [CardInfoType.NUMBER]: (name: string, value: string) => ({
      type: CARD_ACTIONS.CHANGE_CARD_NUMBER,
      payload: { index: Number(name[name.length - 1]), value },
    }),
    [CardInfoType.COMPANY]: (_: string, value: string) => ({
      type: CARD_ACTIONS.CHANGE_CARD_COMPANY,
      payload: value,
    }),
    [CardInfoType.EXPDATE]: (name: string, value: string) => ({
      type: CARD_ACTIONS.CHANGE_CARD_EXP_DATE,
      payload: { key: name.split('-')[1] as 'month' | 'year', value },
    }),
    [CardInfoType.CVC]: (_: string, value: string) => ({
      type: CARD_ACTIONS.CHANGE_CARD_CVC,
      payload: value,
    }),
    [CardInfoType.PASSWORD]: (_: string, value: string) => ({
      type: CARD_ACTIONS.CHANGE_CARD_PASSWORD,
      payload: value,
    }),
  };

  const handleCardInfo = (e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>) => {
    const { name, value } = e.target;

    for (const type of Object.keys(actionMap)) {
      if (name.startsWith(type)) {
        const action = actionMap[type as CardInfoType](name, value);
        cardInfoReducer(action as CardAction);
        return;
      }
    }
  };
  ```

JunilHwang replied:
actionMap과 reducer를 내부에서 결합해서 사용할 수 있을 것 같아요 ㅎㅎ
바깥에서 정제할 필요가 없달까..

어쨌든 말씀하신 것과 비슷하긴 합니다!

reducer를 사용하면 리액트 라이프 사이클과 분리해서 관리할 수 있고, 테스트 코드 작성이 수월해져요!

JunilHwang commented on src/features/cardInfo/hooks/useCardInfo.tsx:
```diff
@@ -68,18 +111,105 @@ export default function useCardInfo() {
+    error,
+    currentSection,
+    isAllSectionsCompleted,
+  };
```
cardInfo의 모든 관심사를 한 곳에서 관리하고 있네요.
useCardInfo 내부의 함수를 여러개의 hook으로 분할한 다음에 조합하는 방식으로 사용하면 어떨까요?

eunsoA replied:
네! 위에서 언급한 대로 수정 및 반영하였습니다! (commit : 8c79f0be05bb608fb2d951e134beed52dccd6962)

JunilHwang reviewed (changes requested):
안녕하세요 클레어! 리뷰가 늦어져서 죄송합니다.. 일단 제가 바로 비행기를 탑승해야 해서 빠르게 현재 보이는 부분들만 리뷰 남겨봅니다 ㅠㅠ
다음 공항 도착하면 이어서 리뷰 해볼게요!

--

지금 보니까 다음 포커스로 넘어가는게 제대로 구현이 안 되어있어요! 일단 이 부분을 우선적으로 신경써서 구현해주시면 좋을 것 같습니다.

그 다음에 useCardInfo 라는 훅이 너무 비대해요!
내부 로직을 여러 개의 hook으로 분할해서 조립하는 방식으로 시도해보면 어떨까요?
꼭 하나의 훅에 모든 내용을 작성하지 않아도 된답니다 ㅎㅎ
훅에서 다른 훅을 호출하는 방식도 고려해보세요.

---

> config 사용에 대한 부분

구조화해서 관리하는게 아직 익숙하지 않은 방식이라서 그런게 크다고 생각해요!
한 번 이렇게도 시도해보고 저렇게도 시도해보면서 클레어만의 기준을 만들어나가면 된답니다 ㅎㅎ

> FSD 아키텍쳐

일단 feature에서 app을 의존하는 부분이 있어요. 의존성이 꼬여버린거죠 ㅎㅎ 이런 부분을 우선적으로 개선해주시면 좋겠어요.

그리고 FSD를 따라하는 방법도 좋고, 개인적으로 FSD에서 어떤 방식으로 관심사를 관리하고자 하는지를  분석해보면 좋다고 생각해요.
가령, feature, entities, shared 같은 계층을 꼭 만들어야 하는건 아니라고 생각해요.
비슷한 유형의 의존성을 묶고, 의존성의 방향이 단방향으로 흐르도록 만들어주기만 해도 충분하지 않나 싶습니다.

> 아무래도 이런 구조이다 보니 features/cardInfo/hooks/useCardInfo가 굉장히 무거워요. 이걸 어떤 식으로 분리해야할지 막막한데, 각 카드 정보에 대해서 상태를 분리해야할지 고민이 돼요.

이는 상세리뷰에 남겼으니 확인 부탁드려요!

> 좀 더 가독성 좋은 코드로 리팩터링하려면 어떤 방향성을 잡으면 좋을지 해먼드의 생각이 궁금해요.

지속적으로 가이드하는 내용이지만, "요구사항이 변경되는 상황"을 가정해보고, 이를 토대로 다른사람이 내 코드를 어떻게 읽고 해석하고 이해할 수 있을지 고민해보면 좋답니다 ㅎㅎ

eunsoA committed:
refactor: cardInfoValidator에서 상수 임포트 경로 수정

eunsoA committed:
style: cardInfoInput CSS에서 border 속성 문법 오류 수정

eunsoA committed:
refactor: features/에서 pages/ 레이어로 페이지 컴포넌트 이동

- FSD 아키텍처를 따라 라우트 단위 페이지 컴포넌트 분리
- cardComplete와 cardRegistration을 features에서 pages로 이동
- 페이지 레이아웃과 기능 관련 컴포넌트의 관심사 분리 개선

eunsoA committed:
refactor: routes.ts를 shared/로 이동하여 단방향 의존성 유지

- app/routes에서 shared/config/routes로 이동
- pages에서 app에 대한 의존성 제거

eunsoA committed:
refactor: getCardCompanyKey 함수의 타입 정의 개선

- 리뷰어 리뷰 반영 (https://github.com/woowacourse/react-payments/pull/475#discussion_r2065347738)

eunsoA committed:
refactor: 카드 정보 핸들러 함수 이름 변경

- cardNumber, cardCompany, cardExpDate, cardCVC, cardPassword 상태 변경 핸들러 함수의 이름을 handler- 에서 change-로 변경

eunsoA committed:
style: 카드 등록 페이지 CSS에서 최대 너비(max-width) 속성 추가

eunsoA committed:
refactor(useCardInfo): 카드 정보 입력 관련 함수들을 역할에 따라 명확히 분리

eunsoA committed:
refactor: 카드 정보 타입(CardInfo) 정의 추가 및 경로 수정

eunsoA committed:
refactor(useCardInfo): 카드 정보 로직 분리

- 공통 인터페이스로 공통 타입 정의
- useCardNumber, useCardCompany, useCardExpDate, useCardCVC, useCardPassword: 각 카드 정보 타입별 처리 로직
- useSectionManager: 섹션 상태 및 네비게이션 관리
- useErrorHandler: 에러 상태 관리

eunsoA commented:
안녕하세요 해먼드!
비행기로 이동하신다니.. 아마도 신혼 여행이겠죠? 너무나도 부럽습니다 .. ✈️✈️
여행으로 바쁘신 와중에도 친절하게 리뷰해주셔서 감사해요. 🙇‍♂️


🌀 FSD 아키텍처 구조로 수정
> 일단 feature에서 app을 의존하는 부분이 있어요. 의존성이 꼬여버린거죠 ㅎㅎ 이런 부분을 우선적으로 개선해주시면 좋겠어요. 
> 그리고 FSD를 따라하는 방법도 좋고, 개인적으로 FSD에서 어떤 방식으로 관심사를 관리하고자 하는지를 분석해보면 좋다고 생각해요. > 가령, feature, entities, shared 같은 계층을 꼭 만들어야 하는건 아니라고 생각해요. > 비슷한 유형의 의존성을 묶고, 의존성의 방향이 단방향으로 흐르도록 만들어주기만 해도 충분하지 않나 싶습니다.

말씀해주신 것 처럼 의존성의 방향이 단방향으로 흐르지 않던 문제들이 있더군요!

  - FSD 아키텍처의 기본 레이어 
      - app: 앱 초기화, 라우팅, 글로벌 설정
      - processes (선택적): 여러 feature를 조합한 상위 프로세스
      - pages (선택적): 라우팅 단위 화면
      - features: 특정 기능(유저 액션 중심)의 UI/로직
      - entities: 비즈니스 도메인 모델과 관련된 로직/데이터
      - shared: 재사용 가능한 UI, 유틸, 상수 등
      👉 의존성 방향: shared → entities → features → processes → pages → app

위의 기본적인 FSD 아키텍처 구조에서 저는 아래의 순서로 의존성 방향을 잡았어요.
| Shared -> entities -> features -> pages -> app 
Feature/ 에서 app/ 을 의존하는 문제도 해결했습니다!

🌀 useCardInfo 분리
- `useCardNumber`, `useCardCompany`, `useCardExpDate`, `useCardCVC`, `useCardPassword`: 각 카드 정보 타입별 처리 로직
- `useSectionManager`: 섹션 상태 및 네비게이션 관리
- `useErrorHandler`: 에러 상태 관리
- `useCardInfo` : 모든 개별 훅을 조합하여 통합된 카드 정보 관리

추가로, 리뷰주셨던 내용도 모두 반영했어요.

결혼식부터 신혼여행까지 바쁘셨을텐데 이번 미션 해먼드 덕분에 많이 배웠습니다! :]
오늘도 좋은 하루 보내세요 🍀

JunilHwang commented on src/features/cardInfo/hooks/useCardCompany.tsx:
```diff
@@ -0,0 +1,29 @@
+  sectionNavigation: SectionNavigation;
+}
+
+export function useCardCompany({ sectionState, sectionNavigation }: UseCardCompanyProps) {
```
sectionState 는 지금 안 쓰이고 있네요!?

JunilHwang commented on src/features/cardInfo/hooks/useCardCompany.tsx:
```diff
@@ -0,0 +1,29 @@
+
+  return {
+    cardCompany,
+    changeCardCompany,
```
이렇게 쓰이는 모습은 어떨까요!?

```tsx
const cardCompany = useCardCompany({ ... });

cardCompany.value;
cardCompany.update('newValue');
```

보통 훅을 통해서 반환하는 값을 항상 해체해서 쓰곤 하는데,
지금처럼 다양한 값을 다루는 경우에는 오히려 해체해서 사용하는게 번잡하게 느껴질 수 있다고 생각해요.
그렇다면 지금 제시해드린 것 처럼 namespace 로 묶어서 사용하는 방법은 어떨까 싶습니다!

이걸 다른 이름으로 사용하고 싶다면 사용자가 destructuring 할 때 alias 를 지정하는 방법도 있으니까요 ㅎㅎ

JunilHwang commented on src/features/cardInfo/hooks/useCardPassword.tsx:
```diff
@@ -0,0 +1,53 @@
+  const changeCardPassword = (value: string) => {
+    updateCardPassword(value);
+
+    const isValid = validatePassword(value);
```
위에서는 isPasswordComplete 이 함수로 쓰이고, 여기서는 boolean으로 쓰이고 있어요.
그리고 validatePassword 라는 함수가 반환하는 값도 boolean 이네요..!
여기서도 isValidPassword 처럼 쓰여야 하지 않을까요!?

네이밍 규칙이 뒤섞여 있다고 느껴집니다 ㅎㅎ

JunilHwang commented on src/features/cardInfo/hooks/useCardPassword.tsx:
```diff
@@ -0,0 +1,53 @@
+
+    const isValid = validatePassword(value);
+
+    if (isValid && isPasswordComplete(value)) {
```
이런 상황에서는... isValid 보단 `validPassword` 처럼 쓰이면 어떨까 싶어요!

JunilHwang commented on src/features/cardInfo/hooks/useCardPassword.tsx:
```diff
@@ -0,0 +1,53 @@
+    if (!configItem || !configItem.validator) return false;
+
+    const [errorIndex, errorMessage] = configItem.validator(password);
+    errorState.setError(
```
`errorState.setError` 는  중복 표현이라고 느껴져요!

- errorState.set
- errorState.change
- errorState.update

요렇게 쓰이면 어떨까 싶네요!?

JunilHwang commented on src/features/cardInfo/hooks/useCardCVC.tsx:
```diff
@@ -0,0 +1,49 @@
+  const changeCardCVC = (value: string) => {
+    updateCardCVC(value);
+
+    const isValid = validateCVC(value);
```
전체적으로 비슷한 구조인 것 같네요!?

한 번에 묶어서 표현해줄 수 있는 방법이 있지 않을까요~?

아니면 요렇게 쓰인다거나
```tsx
const useCardCVC = () => useCardInput(...);
```

useCardInput 이라는걸 만들고, 이를 토대로 각각의 훅을 생성하는 방식은 어떨까 싶어요!

JunilHwang commented on src/features/cardInfo/hooks/useErrorHandler.tsx:
```diff
@@ -0,0 +1,38 @@
+
+  return {
+    error,
+    getErrorState,
```
어차피 error을 넘겨주고 있어서
getErrorState 대신에 setError 을 넘겨주면 되지않을까요!?

JunilHwang commented on src/features/cardInfo/hooks/useSectionManager.tsx:
```diff
@@ -0,0 +1,54 @@
+
+  const getSectionState = (sectionType: CardInfoType) => ({
+    isCompleted: completedSections[sectionType],
+    setCompleted: (isCompleted: boolean) => setSectionState(sectionType, isCompleted),
```
흠.. 전체적으로 get이 setter 까지 반환하는 구조로 되어있네요.

그렇다면 getSectionState 라는 이름보다 더 어울리는 이름이 필요하지 않을까요!?

JunilHwang commented on src/pages/cardComplete/CardComplete.tsx:
```diff
@@ -0,0 +1,34 @@
+
+  const handleConfirm = () => {
+    navigate(ROUTES.CARD_REGISTRATION);
+  };
```
함수의 내용을 보면 결국 `CARD_REGISTRATION` 페이지로 이동하는 일을 하고 있어요.
이벤트를 받아서 처리하는 것도 아니기 때문에, 이럴 때는

`goToRegistration` 처럼 표현해주면 어떨까요?

JunilHwang commented on src/features/cardInfo/hooks/useCardInfo.tsx:
```diff
@@ -1,85 +1,92 @@
-  return { cardInfo, setCardInfo, handleCardInfoChange, error };
+  return {
+    cardInfo,
+    handleCardInfo,
```
`handleCardInfo` 가 무엇을 의미하는걸까요!?
차라리 `handleChange` 처럼 표현해주면 어떨까 싶어요!

JunilHwang reviewed (approved):
안녕하세요 클레어!
리팩토링 잘 진행해주셨네요 ㅎㅎ

현재 설계에 대해서는 큰 이견 없습니다!
다만 네이밍 규칙 관련해서 몇 가지 리뷰 남겼으니 참고해주세요~!

페이먼츠 미션은 여기서 마무리하도록 하겠습니다.
부족한 리뷰어였으나 클레어에게 조금이나마 도움이 되었길 바랍니다!
고생하셨어요 클레어~! 나머지 미션도 무사완주 하기를 바랍니다!
